File: Src/Icm.Core/Functions/IKeyedMathFunction.vb
Imports Icm.Collections.Generic.StructKeyStructValue

Namespace Icm.Functions
    Public Interface IKeyedMathFunction(Of TX As {Structure, IComparable(Of TX)}, TY As {Structure, IComparable(Of TY)})
        Inherits IMathFunction(Of TX, TY)

        Function Range(ByVal rangeStart As TX, ByVal rangeEnd As TX, ByVal includeExtremes As Boolean) As RangeIterator(Of TX, TY)
        Function RangeFrom(ByVal rangeStart As TX, ByVal includeExtremes As Boolean) As RangeIterator(Of TX, TY)
        Function RangeTo(ByVal rangeEnd As TX, ByVal includeExtremes As Boolean) As RangeIterator(Of TX, TY)
        Function TotalRange(ByVal includeExtremes As Boolean) As RangeIterator(Of TX, TY)
        Function PreviousOrLst(ByVal d As TX) As TX?
        ReadOnly Property KeyStore() As ISortedCollection(Of TX, TY)
    End Interface
End Namespace

File: Src/Icm.Core/Functions/IKeyedMathFunction.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Icm.Collections.Generic.StructKeyStructValue;

namespace Icm.Functions
{
    public interface IKeyedMathFunction<TX, TY> : IMathFunction<TX, TY> where TX : struct, IComparable<TX> where TY : struct, IComparable<TY>
    {

        RangeIterator<TX, TY> Range(TX rangeStart, TX rangeEnd, bool includeExtremes);
        RangeIterator<TX, TY> RangeFrom(TX rangeStart, bool includeExtremes);
        RangeIterator<TX, TY> RangeTo(TX rangeEnd, bool includeExtremes);
        RangeIterator<TX, TY> TotalRange(bool includeExtremes);
        TX? PreviousOrLst(TX d);
        ISortedCollection<TX, TY> KeyStore { get; }
    }
}
